src/diagnostics.rs

```diff
@@ -0,0 +1,156 @@
+/*! Diagnostic reporting functionality
+
+When enabled with the `diagnostics` feature (default) this module provides automated install success/failure reporting to an endpoint.
+
+That endpoint can be a URL such as `https://our.project.org/nix-installer/diagnostics` or `file:///home/$USER/diagnostic.json` which recieves a [`DiagnosticReport`] in JSON format.
+*/
+
+use std::time::Duration;
+
+use os_release::OsRelease;
+use reqwest::Url;
+
+/// The static of an action attempt
+#[derive(Debug, serde::Deserialize, serde::Serialize, Clone)]
+pub enum DiagnosticStatus {
+    Cancelled,
+    Success,
+    /// This includes the [`strum::IntoStaticStr`] representation of the error, we take special care not to include parameters of the error (which may include secrets)
+    Failure(String),
```

Comment:
I wonder, could we easily include which step failed as well...? Knowing it was a permissions error would be more useful if it also said "CreateFile".

Reply:
Yeah... It's something I'd like to improve. I think we have some work to do on error reporting in general to be honest. I'd like to wrap that into the later work.

Reply:
Sounds good to me. I opened https://github.com/DeterminateSystems/nix-installer/issues/268 for that.